Refactors MultiMatchQueryBuilder and Parser

alexksikes opened the pull request:
Relates to #10217

This PR is against the query-refactoring branch.

javanna commented:
@cbuescher can you review this once you are done with the match query please?

cbuescher commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -313,6 +313,14 @@ public final Boolean readOptionalBoolean() throws IOException {
+        }
+        return null;
+    }
+
```
I think `Object readGenericValue()` works for `Float` and we don't need this.

cbuescher commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -201,6 +201,15 @@ public void writeOptionalVInt(@Nullable Integer integer) throws IOException {
+            writeFloat(aFloat);
+        }
+    }
+
```
Same here, `writeGenericValue` seems to do the same.

alexksikes replied:
`writeGenericValue` also works on Strings etc .. and also handles optionals. This is more for readable. Question: should we use the generic method all the time, or its writeX counter part?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -295,27 +350,40 @@ public MultiMatchQueryBuilder cutoffFrequency(float cutoff) {
+    @Deprecated
     public MultiMatchQueryBuilder zeroTermsQuery(MatchQueryBuilder.ZeroTermsQuery zeroTermsQuery) {
-        this.zeroTermsQuery = zeroTermsQuery;
+        this.zeroTermsQuery = ZeroTermsQuery.valueOf(zeroTermsQuery.name().toUpperCase(Locale.ROOT));
```
fyi, I removed the ZeroTermsQuery class from MatchQueryBuilder entirely in #13402, moved everything plus serialization to #13402, so we might remove this setter and only have one.

alexksikes replied:
sure sounds good

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryParser.java:
```diff
@@ -50,31 +45,41 @@ public MultiMatchQueryParser() {
+        Float tieBreaker = null;
+        Float cutoffFrequency = null;
+        Boolean lenient = null;
+        MatchQuery.ZeroTermsQuery zeroTermsQuery = null;
```
We can have ZeroTermsQuery.NONE here as a default, thats what MatchQuery uses. Then some null-checks can be removed later. I added a default constant to MatchQuery here: https://github.com/elastic/elasticsearch/pull/13402/files#diff-943345b14010c833dc54a52b10a28489R128

alexksikes replied:
Sure let's make it consistent with MatchQuery

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -179,7 +193,7 @@ public MultiMatchQueryBuilder field(String field, float boost) {
      * Sets the type of the text query.
      */
-    public MultiMatchQueryBuilder type(MultiMatchQueryBuilder.Type type) {
+    public MultiMatchQueryBuilder type(Type type) {
```
Wonder if it makes sense to rename the inner `Type` here to something more specific to not get it confused with the three `MatchQuery.Type`s that it's using internally.

alexksikes replied:
I agree Type is very used name space, so MatchQuery.Type makes it more explicit.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -218,10 +232,20 @@ public MultiMatchQueryBuilder slop(int slop) {
+     * Set the phrase slop if evaluated to a phrase query type.
+     */
+    public MultiMatchQueryBuilder slop(Integer slop) {
+        this.slop = slop;
```
While refactoring MatchQ.B. I change slop to `int` and using default if not set. See https://github.com/elastic/elasticsearch/pull/13402#discussion-diff-38959205R137 for discussion.

alexksikes replied:
I agree we should actually set defaults here, and find these defaults in the actual MatchQuery.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -230,6 +254,11 @@ public MultiMatchQueryBuilder prefixLength(int prefixLength) {
     }
 
+    public MultiMatchQueryBuilder prefixLength(Integer prefixLength) {
+        this.prefixLength = prefixLength;
```
Same here, we can change to primitive `int` and use default values for this and some more of the fuzzy options. This will also simplify doXContent, can also be used in parser then.

alexksikes replied:
Note if we use defaults instead of null, then we have to always write them and render them. From a methodology perspective I think we should always try to find the defaults in the actual query whenever they exist and not use null.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -373,4 +434,196 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        out.writeOptionalFloat(tieBreaker);
+        out.writeOptionalBoolean(lenient);
+        out.writeOptionalFloat(cutoffFrequency);
+        if (zeroTermsQuery != null) {
```
fyi, I added serialization to the enum I moved to MatchQuery (#13402) maybe we can reuse this.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -373,4 +434,196 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        out.writeGenericValue(fieldsBoosts);
+        if (type != null) {
+            out.writeBoolean(true);
+            out.writeString(type.name().toUpperCase(Locale.ROOT));
```
Can we make MultiMatchQuery.Type a Writable similar to Operator etc...? Relying on serializing an ordinal feels safer to me than using the names.

alexksikes replied:
Yeah we already had this discussion with the Operator Enum. I think this is a bit too over-engineering the matter, but I do like what @s1monw recently did here with serializing Enums.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -373,4 +434,196 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+            }
+            multiMatchQuery.setAnalyzer(analyzer);
+        }
+        if (slop != null) {
```
Some of these checks can potentially be removed after changing internal field to primitve types / using defaults.

alexksikes replied:
Yes in this case we would just be setting the defaults, which are already set in the internal query. Somehow for me, not setting something really means use defaults.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    protected MultiMatchQueryBuilder doCreateTestQueryBuilder() {
+        //NO COMMIT: we should test multiple fields, but some are incompatible with the provided value
```
Could we maybe extend the setup for this special test to add some more fields with same field type? Like, for example, choose a random type and then add some more fields with same mappings, then use it here for the query as well? Not sure about the details though, sorry.

javanna replied:
in the query_string PR I did something very basic, added a second string field to the base query test case setup, have a look at #13284

alexksikes replied:
@javanna I saw what you did, looks good. What I'll do is simply add multiple fields of the same random type.

cbuescher commented:
@alexksikes did a first round of reviews. Since this query relies much in MatchQueryBuilder which is currently WIP in #13402, there is a bit of movement here. Also, in working on the match query I first went with having many fields as optional objects, but after discussion today changed many to using primitive types plus defaults. Maybe this applies in other places for this PR as well.

alexksikes commented:
@cbuescher I rebased and addressed all the comments. Thanks for the review.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -295,82 +411,238 @@ public MultiMatchQueryBuilder cutoffFrequency(float cutoff) {
+    @Override
+    protected void setFinalBoost(Query query) {
+        query.setBoost(boost * query.getBoost());
+    }
```
Took me some time to get this overwrite, maybe some javadocs would help here.

javanna replied:
I think the base method in AbstractQueryBuilder has already extensive javadocs, maybe we can add a simple comment here explaining that the query might get a boost coming out of field boosts parsing, which we need to combine with the boost assigned to the query as part of the query dsl.

That said, I don't see this same behaviour in master, it feels like it is the right behaviour especially looking at what query_string does, but on master we actually replace the parsed boost with the boost assigned to the query in the query DSL. 

Maybe open a separate issue and mark it fixed together with this PR so we at least keep track of this small change?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -295,82 +411,238 @@ public MultiMatchQueryBuilder cutoffFrequency(float cutoff) {
-        return NAME;
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (value == null) {
```
Could this go to the constructor now?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryParser.java:
```diff
@@ -49,31 +45,41 @@ public MultiMatchQueryParser() {
-        Float tieBreaker = null;
-        MultiMatchQueryBuilder.Type type = null;
-        MultiMatchQuery multiMatchQuery = new MultiMatchQuery(context);
+        Map<String, Float> fieldsBoosts = new HashMap<>();
```
This was a TreeMap in the builder, just curious if this is different on purpose.

alexksikes replied:
The key, values will be put into the builder Treemap, so it really does not matter as long as it is a Map.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java:
```diff
@@ -0,0 +1,185 @@
+
+    @Override
+    protected void doAssertLuceneQuery(MultiMatchQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
+
```
Maybe put some comment why we don't to further checks here or if there are more specialized test below? Otherwise one might wonder why this is here.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java:
```diff
@@ -0,0 +1,185 @@
+        assertThat(bQuery.clauses().size(), equalTo(2));
+        assertThat(assertBooleanSubQuery(query, TermQuery.class, 0).getTerm(), equalTo(new Term(STRING_FIELD_NAME, "test")));
+        assertThat(assertBooleanSubQuery(query, TermQuery.class, 1).getTerm(), equalTo(new Term(STRING_FIELD_NAME_2, "test")));
+    }
```
Nice, I like it to have some more controlled, non randomized tests here for basic behaviour.

cbuescher commented:
@alexksikes left a few comments, very minor and mostly open questions on my side. Not sure if someone else should take a final look, otherwise looks good to me.

alexksikes commented:
@cbuescher I addressed all the comments. In some specific cases MatchQueryBuilder (and so does MultiMatch) fails on date fields, which is untested yet on MatchQueryBuilder. I'll open a PR for this. Thanks for the review.

cbuescher commented:
Had a look at the last changes, LGTM
